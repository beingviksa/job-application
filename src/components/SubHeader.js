const SubHeader = () => {
  return (
    <div className="description">
      we will use the data we collect here to better target your audience.
    </div>
  );
};

export default SubHeader;
