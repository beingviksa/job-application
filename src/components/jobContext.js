import { createContext, useContext, useState } from "react";

export const JobContext = createContext();

export const JobContextProvider = ({ children }) => {
  const [formData, setFormData] = useState({});
  const [formErrors, setFormErrors] = useState({});

  return (
    <JobContext.Provider
      value={{
        formData,
        setFormData,
        formErrors,
        setFormErrors,
      }}>
      {children}
    </JobContext.Provider>
  );
};

export const useJobContext = () => {
  return useContext(JobContext);
};
