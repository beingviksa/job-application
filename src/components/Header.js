import SubHeader from "./SubHeader";

const Header = ({
  iconSymbol,
  iconBgColor,
  headerContent,
  showSubHeader = false,
}) => {
  return (
    <header>
      <div className="job-title">
        <p className="form-title-icon" style={{ backgroundColor: iconBgColor }}>
          {iconSymbol}
        </p>
        <h1>{headerContent}</h1>
      </div>

      {showSubHeader && <SubHeader />}
    </header>
  );
};

export default Header;
