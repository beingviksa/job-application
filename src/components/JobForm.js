import { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import Button from "./UI/Button";
import Checkbox from "./UI/Checkbox";
import { educationLevels } from "../utills/educationLevels";
import Header from "./Header";
import { useJobContext } from "./jobContext";
import ErrorText from "./UI/Error";

const JobForm = () => {
  const navigate = useNavigate();
  const { setFormData, formData, formErrors, setFormErrors } = useJobContext();

  const initialData = {
    experience: formData.experience || null,
    minWorkingHours: formData.workingHours?.min,
    maxWorkingHours: formData.workingHours?.max,
    educationData: formData.educationData,
  };

  const [experience, setExperience] = useState(initialData.experience);
  const [educationData, setEducationData] = useState(
    initialData.educationData || educationLevels
  );
  const [workingHours, setWorkingHours] = useState({
    min: initialData.minWorkingHours || 20,
    max: initialData.maxWorkingHours || 50,
  });

  const handleChangeQualification = (id) => {
    setEducationData((prev) => {
      return prev.map((item) => {
        if (item.id === id) {
          return { ...item, checked: !item.checked };
        } else {
          return { ...item };
        }
      });
    });
  };

  const handleChangeExperience = (e) => {
    setExperience(e.target.value);
  };

  const handleWorkingHourChange = (e) => {
    setWorkingHours({
      ...workingHours,
      [e.target.name]: e.target.value,
    });
  };

  const validateForm = (values) => {
    const errors = {};
    if (!values?.experience) {
      errors.experience = "The field is required";
    }

    if (values.educationData.every((i) => !i.checked)) {
      errors.educationData = "Please select the qualification";
    }

    if (!values?.workingHours?.min || !values?.workingHours?.max) {
      errors.workingHours = "Please add working hours";
    }

    setFormErrors(errors);

    return errors;
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    const data = {
      experience,
      educationData,
      workingHours,
    };

    setFormData({ ...data });
    const errors = validateForm({ ...data });

    if (Object.keys(errors).length === 0) {
      navigate("/overview");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <div className="job-form">
        <Header
          iconSymbol="1"
          iconBgColor="violet"
          headerContent="Job criteria"
          showSubHeader
        />
        <div className="input-field number-input">
          <label className="field-title">
            A minimum No. years of experience
          </label>
          <input
            type="number"
            value={experience}
            onChange={handleChangeExperience}
            placeholder="e.g 5+"
            min="1"
            max="40"
          />
          {formErrors?.experience && (
            <ErrorText error={formErrors.experience} />
          )}
        </div>
        <div className="input-field">
          <label className="field-title">Level of education</label>
          {educationData.map(({ id, name, value }, index) => {
            return (
              <Checkbox
                index={index}
                name={name}
                value={value}
                id={id}
                checkedState={educationData}
                handleOnChange={() => handleChangeQualification(id)}
                key={index}
              />
            );
          })}
          {formErrors?.educationData && (
            <ErrorText error={formErrors.educationData} />
          )}
        </div>
        <div className="input-field">
          <p className="field-title">No. of working hours (per week)</p>
          <div className="working-hours-field number-input">
            <label>
              Min.{" "}
              <input
                type="number"
                name="min"
                value={workingHours.min}
                min="20"
                max="80"
                onChange={handleWorkingHourChange}
              />
            </label>
            <label>
              Max.{" "}
              <input
                type="number"
                name="max"
                value={workingHours.max}
                min="40"
                max="80"
                onChange={handleWorkingHourChange}
              />
            </label>
          </div>
          {formErrors?.workingHours && (
            <ErrorText error={formErrors.workingHours} />
          )}
        </div>
        <Button type="submit">Save</Button>
      </div>
    </form>
  );
};

export default JobForm;
