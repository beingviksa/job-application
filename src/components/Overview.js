import { useNavigate } from "react-router-dom";

import Header from "./Header";
import { useJobContext } from "./jobContext";
import Button from "./UI/Button";
import Tag from "./UI/Tag";

const Overview = () => {
  const navigate = useNavigate();
  const {
    formData: { experience, workingHours, educationData },
  } = useJobContext();
  const { min, max } = workingHours;

  const handleClick = () => {
    navigate("/");
  };

  return (
    <div className="job-form overview-page">
      <div className="overview-header">
        <Header
          iconSymbol="&#10003;"
          iconBgColor="green"
          headerContent="Job criteria"
        />
        <Button handleClick={handleClick}>Edit</Button>
      </div>

      <div className="overview-content">
        <p className="overview-content__title">
          A minimum No. years of experience:{" "}
          <span className="overview-content__response">{experience}</span>
        </p>
        <p className="overview-content__title">
          No. of working hours (per week):
          <span className="overview-content__response">
            &nbsp;{`${min}-${max} hours`}
          </span>
        </p>
        <div className="overview-content__education">
          <h3>Level of education</h3>
          <div className="tags">
            {educationData
              .filter((ed) => ed.checked)
              .map((data) => {
                return <Tag content={data.value} key={data.index} />;
              })}
          </div>
        </div>
      </div>
    </div>
  );
};

export default Overview;
