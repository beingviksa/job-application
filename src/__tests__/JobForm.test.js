import { fireEvent, render, waitFor } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import JobForm from "../components/JobForm";
import { JobContext } from "../components/jobContext";

describe("<Job Form/>", () => {
  const contextData = {
    formData: {
      experience: "4",
      educationData: [
        {
          id: 1,
          name: "bachelor",
          value: "Bachelor / Graduate",
          checked: true,
        },
        {
          id: 2,
          name: "highschool",
          value: "GCSE / A-Level / Highschool / GED",
          checked: false,
        },
        {
          id: 3,
          name: "master",
          value: "Master / Post-Graduate /PhD",
          checked: false,
        },
        {
          id: 4,
          name: "diploma",
          value: "Vocational / Diploma / Associates degree",
          checked: false,
        },
      ],
      workingHours: {
        min: 20,
        max: 50,
      },
    },
    setFormData: jest.fn(),
    setFormErrors: jest.fn(),
    formErrors: {},
  };

  it("render the JobForm Component", async () => {
    const { container } = render(
      <JobContext.Provider value={contextData}>
        <Router>
          <JobForm />
        </Router>
      </JobContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("set formData on submit button click", async () => {
    const { getByTestId } = render(
      <JobContext.Provider value={contextData}>
        <Router>
          <JobForm />
        </Router>
      </JobContext.Provider>
    );

    const submitButton = getByTestId("submit-button");
    fireEvent.click(submitButton);

    await waitFor(() => {
      expect(contextData.setFormData).toBeCalled();
    });
  });

  it("show error message onsubmit for empty field", async () => {
    const contextData = {
      formData: {
        experience: null,
        educationData: [
          {
            id: 1,
            name: "bachelor",
            value: "Bachelor / Graduate",
            checked: false,
          },
          {
            id: 2,
            name: "highschool",
            value: "GCSE / A-Level / Highschool / GED",
            checked: false,
          },
          {
            id: 3,
            name: "master",
            value: "Master / Post-Graduate /PhD",
            checked: false,
          },
          {
            id: 4,
            name: "diploma",
            value: "Vocational / Diploma / Associates degree",
            checked: false,
          },
        ],
        workingHours: {
          min: 20,
          max: 50,
        },
      },
      setFormData: jest.fn(),
      setFormErrors: jest.fn(),
    };

    const { getByTestId } = render(
      <JobContext.Provider value={contextData}>
        <Router>
          <JobForm />
        </Router>
      </JobContext.Provider>
    );

    const submitButton = getByTestId("submit-button");
    fireEvent.click(submitButton);

    expect(contextData.setFormErrors).toBeCalledWith({
      educationData: "Please select the qualification",
      experience: "The field is required",
    });
  });
});
