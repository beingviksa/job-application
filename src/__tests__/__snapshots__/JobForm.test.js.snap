// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Job Form/> render the JobForm Component 1`] = `
<div>
  <form>
    <div
      class="job-form"
    >
      <header>
        <div
          class="job-title"
        >
          <p
            class="form-title-icon"
            style="background-color: violet;"
          >
            1
          </p>
          <h1>
            Job criteria
          </h1>
        </div>
        <div
          class="description"
        >
          we will use the data we collect here to better target your audience.
        </div>
      </header>
      <div
        class="input-field number-input"
      >
        <label
          class="field-title"
        >
          A minimum No. years of experience
        </label>
        <input
          max="40"
          min="1"
          placeholder="e.g 5+"
          type="number"
          value="4"
        />
      </div>
      <div
        class="input-field"
      >
        <label
          class="field-title"
        >
          Level of education
        </label>
        <div
          class="checkbox"
        >
          <input
            checked=""
            id="custom-checkbox-0"
            name="bachelor"
            type="checkbox"
            value="Bachelor / Graduate"
          />
          <label
            for="custom-checkbox-0"
          >
            Bachelor / Graduate
          </label>
        </div>
        <div
          class="checkbox"
        >
          <input
            id="custom-checkbox-1"
            name="highschool"
            type="checkbox"
            value="GCSE / A-Level / Highschool / GED"
          />
          <label
            for="custom-checkbox-1"
          >
            GCSE / A-Level / Highschool / GED
          </label>
        </div>
        <div
          class="checkbox"
        >
          <input
            id="custom-checkbox-2"
            name="master"
            type="checkbox"
            value="Master / Post-Graduate /PhD"
          />
          <label
            for="custom-checkbox-2"
          >
            Master / Post-Graduate /PhD
          </label>
        </div>
        <div
          class="checkbox"
        >
          <input
            id="custom-checkbox-3"
            name="diploma"
            type="checkbox"
            value="Vocational / Diploma / Associates degree"
          />
          <label
            for="custom-checkbox-3"
          >
            Vocational / Diploma / Associates degree
          </label>
        </div>
      </div>
      <div
        class="input-field"
      >
        <p
          class="field-title"
        >
          No. of working hours (per week)
        </p>
        <div
          class="working-hours-field number-input"
        >
          <label>
            Min.
             
            <input
              max="80"
              min="20"
              name="min"
              type="number"
              value="20"
            />
          </label>
          <label>
            Max.
             
            <input
              max="80"
              min="40"
              name="max"
              type="number"
              value="50"
            />
          </label>
        </div>
      </div>
      <button
        data-testid="submit-button"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
